### PR TITLE
Add enum macro

### PR DESCRIPTION
**Pull request by tenedor**
Here's a macro that uses the `NatOtherThan` approach and includes the full proofs.

I think it makes sense to include pretty-print names as part of the enum declarations, at least for the constants where we're customizing pretty print names.

It's not clear to me how important it is to support enum terms that match a range of values other than the single default case. Clearly this pattern comes up somewhat regularly in the ELFSage constants, but it's not clear to me what the practical trade is on supporting this pattern explicitly. My hunch is to not support this for now but to always name the enum parameter explicitly, e.g. `(n : Nat)`, so that the syntax is consistent if this gets generalized down the road, but I don't have a strong opinion here.

Are there cases in the ELFSage constants that would be valuable to represent as enumerations over `Fin n` instead of `Nat`?

I gave examples for how these generalizations could look syntactically in the comments below my changes.

### CODE REVIEW

File: ELFSage/test.lean [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  let enumCases := cases.pop
+  let (names, values, comments) := splitArray enumCases
+  let (defaultName, defaultValue, defaultComment) := split defaultCase
+  let valuesList: TSyntax `term := Lean.quote values.toList
+  let rawNames := names.map (·.getId)
+  let strNames := rawNames.map toStrLit
+  let prefixedNames := rawNames.map (eIdent ·)
+
+  if ¬ n == defaultValue then
+    -- TODO is there a better way to error? Panicking results in a terrible authoring experience.
```
Comment (gleachkr) [inline ELFSage/test.lean]: I think Macro.throwErrorAt might do the trick: https://leanprover-community.github.io/mathlib4_docs/Init/Prelude.html#Lean.Macro.throwErrorAt

File: ELFSage/test.lean [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+private def splitArray (cases : TSyntaxArray `enum_case) : (Array Ident) × (Array Term) × (Array (Lean.TSyntax `Lean.Parser.Command.docComment)) :=
+  cases.foldl (init := (#[], #[], #[])) fun (names, values, comments) => fun (case) =>
+    let (name, value, comment) := split case
+    (names.push name, values.push value, comments.push comment)
+
+/--
+The `enum` macro creates named constants for each case of an enumeration type.
+A default case is created for any value not explicitly enumerated.
+
+For now, the enum value type must be `Nat`. This may be generalized to support
```
Comment (gleachkr) [inline ELFSage/test.lean]: Is this generalization functionality already available? I don't see anything that specifically depends on `Nat`, and it seems like it's enough to write `enum E (Fin 5) where …` to get an enum over `Fin 5`.

File: ELFSage/test.lean (hunk 1 above)
Comment (tenedor) [reply ELFSage/test.lean]: Much better!

File: ELFSage/test.lean (hunk 2 above)
Comment (tenedor) [reply ELFSage/test.lean]: Whoops - that comment was leftover from when I was using `NatOtherThan`

**Comment by tenedor** [comment]
@gleachkr this works now!

Here are examples:

```lean 4
-- can mix literal, guarded, and match-all values. comments are optional
enum E (n: Nat)
  /-- one -/
  | one => 1
  | small => n, 2 ≤ n ∧ n < 4
  /-- four -/
  | four => 4
  /-- other -/
  | other => n

#eval E.four
#eval E.ofVal 4
#eval E.small 2 (by decide)
#eval (E.other 5 (by decide)).val
#eval (E.ofVal 5).val

-- proves round-trip identities for both ofVal ⬝ val and val ⬝ ofVal
#check E.ofVal_val
#check E.val_ofVal

-- works with types other than Nat
enum FinEnum (n: Fin 3)
  | zero => 0
  | one => 1
  | two => 2

#eval FinEnum.zero.val
#eval FinEnum.ofVal 0

-- adds a default case if it isn't given explicitly
enum ImplicitDefault (n: Nat)
  | one => 1
  | two => 2

#eval ImplicitDefault.default 5 (by decide)
#eval ImplicitDefault.ofVal 5
```

Here's (approximately) what the macro generates for the first example:

```lean 4
inductive E
  | one
  | small (n : Nat) (h: (2 ≤ n ∧ n < 4) ∧ ¬(n = 1))
  | four
  | other (n : Nat) (h: ¬(n = 1) ∧ ¬(2 ≤ n ∧ n < 4) ∧ ¬ (n = 4))
  deriving BEq, Inhabited

instance : Repr E where
  reprPrec := fun e _ => match e with
    | .one => "one"
    | .small n _ => s!"small {n}"
    | .four => "four"
    | .other n _ => s!"other {n}"

def E.val : E → Nat
  | E.one => 1
  | E.small n _ => n
  | E.four => 4
  | E.other n _ => n

def E.ofVal : Nat → E := fun n =>
  if _ : n = 1 then .one else
  if _ : 2 ≤ n ∧ n < 4 then .small n (by omega) else
  if _ : n = 4 then .four else
  if _ : true then .other n (by omega) else
  unreachable!

theorem E.ofVal_val (e : E) : E.ofVal (E.val e) = e := by
  cases e <;> try rfl
  all_goals rename _ => h; simp [E.val, E.ofVal, h]

theorem E.val_ofVal (n : Nat) : E.val (E.ofVal n) = n := by
  unfold E.ofVal
  repeat split; simp only [*, E.val]
  simp only [*, E.val]
  contradiction
```

**Comment by gleachkr** [comment]
This looks awesome. How do you feel, is it ready to go? If so, maybe a good home would be `ELFSage/Macros/enum.lean`?

**Comment by tenedor** [comment]
@gleachkr great, glad you like it! I think it's ready to go, yep. I moved the file to `ELFSage/Macros/Enum.lean`. Let me know if you were deliberately recommending lowercase `enum.lean`.

For the llvm-readobj tests, we'll benefit from extending this macro to let an enum value's toString use a custom value that differs from the enum case name. But that will be a modest addition, and we can deal with that when we get to it.

For now, the next step would be to port some constants into this structure and confirm it's working and feels good.

**Pull request merged**